libImaging/Storage.c

```diff
@@ -265,50 +264,178 @@ ImagingDelete(Imaging im)
 /* Allocate image as an array of line buffers. */
 
+struct ImagingMemoryArena ImagingDefaultArena = {
+    1,                   // alignment
+    1*1024*1024,         // block_size
+    0,                   // blocks_max
+    0,                   // blocks_cached
+    NULL,                // blocks
+    0, 0, 0, 0, 0        // Stats
+};
+
+int
+ImagingMemorySetBlocksMax(ImagingMemoryArena arena, int blocks_max)
+{
+    void *p;
+    /* Free already cached blocks */
+    ImagingMemoryClearCache(arena, blocks_max);
+
+    if (blocks_max == 0 && arena->blocks != NULL) {
+        free(arena->blocks);
+        arena->blocks = NULL;
+    } else if (arena->blocks != NULL) {
+        p = realloc(arena->blocks, sizeof(*arena->blocks) * blocks_max);
+        if ( ! p) {
+            // Leave previous blocks_max value
+            return 0;
+        }
+        arena->blocks = p;
+    } else {
+        arena->blocks = calloc(sizeof(*arena->blocks), blocks_max);
+        if ( ! arena->blocks) {
+            return 0;
+        }
+    }
+    arena->blocks_max = blocks_max;
+
+    return 1;
+}
+
+void
+ImagingMemoryClearCache(ImagingMemoryArena arena, int new_size)
+{
+    while (arena->blocks_cached > new_size) {
+        arena->blocks_cached -= 1;
+        free(arena->blocks[arena->blocks_cached].ptr);
+        arena->stats_freed_blocks += 1;
+    }
+}
+
+ImagingMemoryBlock
+memory_get_block(ImagingMemoryArena arena, int requested_size, int dirty)
+{
+    ImagingMemoryBlock block = {NULL, 0};
+
+    if (arena->blocks_cached > 0) {
+        // Get block from cache
+        arena->blocks_cached -= 1;
+        block = arena->blocks[arena->blocks_cached];
+        // Reallocate if needed
+        if (block.size != requested_size){
+            block.ptr = realloc(block.ptr, requested_size);
+        }
+        if ( ! block.ptr) {
+            // Can't allocate, free prevous pointer (it is still valid)
+            free(arena->blocks[arena->blocks_cached].ptr);
+            arena->stats_freed_blocks += 1;
+            return block;
+        }
+        if ( ! dirty) {
+            memset(block.ptr, 0, requested_size);
+        }
+        arena->stats_reused_blocks += 1;
+        if (block.ptr != arena->blocks[arena->blocks_cached].ptr) {
+            arena->stats_reallocated_blocks += 1;
+        }
+    } else {
+        if (dirty) {
+            block.ptr = malloc(requested_size);
+        } else {
+            block.ptr = calloc(1, requested_size);
+        }
+        arena->stats_allocated_blocks += 1;
+    }
+    block.size = requested_size;
+    return block;
+}
+
+void
+memory_return_block(ImagingMemoryArena arena, ImagingMemoryBlock block)
+{
+    if (arena->blocks_cached < arena->blocks_max)  {
+        // Reduce block size
+        if (block.size > arena->block_size) {
+            block.size = arena->block_size;
+            block.ptr = realloc(block.ptr, arena->block_size);
+        }
+        arena->blocks[arena->blocks_cached] = block;
```

Comment:
This is going to potentially have a race condition if it's not done within the GIL

Reply:
Yes, that is why `ImagingAllocateArray` [isn't thread-safe anymore](https://github.com/python-pillow/Pillow/pull/2738/files#diff-7e3bc1c431db9d44de31c360569407f9L287)

Reply:
Ah. I see that.